[`flake8-builtins`] Make strict module name comparison optional (`A005`)

ntBre opened the pull request:
## Summary

This PR adds the configuration option `lint.flake8-builtins.builtins-strict-checking`, which is used in A005 to determine whether the fully-qualified module name (relative to the project root or source directories) should be checked instead of just the final component as is currently the case.

As discussed in https://github.com/astral-sh/ruff/issues/15399#issuecomment-2587017147, the default value of the new option is `false`, so modules like `utils.logging` from the initial report are no longer flagged by default.

## Test Plan

New A005 test module with the structure reported in #15399.

ntBre commented on crates/ruff_linter/src/rules/flake8_builtins/rules/stdlib_module_shadowing.rs:
```diff
@@ -85,13 +88,26 @@ pub(crate) fn stdlib_module_shadowing(
         return None;
     }
 
+    // all of the modules considered by `is_known_standard_library` are top-level packages, so if
+    // `path` has a parent directory other than `project_root` and any of the `src` directories, it
+    // should not match in non-strict mode
```
Before the last commit, I was doing more processing on the input `path` to build a fully-qualified module name below the package root and `src` directories to pass to `is_known_standard_library`, but then I noticed this and simplified the check to this parent code.

This was inspired by some of the checks in [`implicit_namespace_package.rs`](https://github.com/astral-sh/ruff/blob/main/crates/ruff_linter/src/rules/flake8_no_pep420/rules/implicit_namespace_package.rs#L77), which also checks for shebang lines and PEP 723 scripts. Should we do that there too?

MichaReiser replied:
I don't think we should exclude PEP 723 scripts or scripts with shebang lines that are at the top level and have the same name as a standard library module because it doesn't change that they shadow the built-in module. Or what's your reasoning for wanting to ignore them?

ntBre replied:
I thought it was related to this comment https://github.com/astral-sh/ruff/issues/15399#issuecomment-2582929030, but my main reasoning was just that it was handled specially by the other rule. I think I agree with you that it makes sense *not* to do it though.

MichaReiser commented:
Have you considered implementing the approach mentioned in https://github.com/astral-sh/ruff/issues/15399#issuecomment-2612005957 and if so, why did you decide against it?

ntBre commented:
I could certainly be missing something here, but from my reading of the comment, this implementation (with the bugfix I just added) should cover that case. Possibly this should be added as a test, but I just added this directrory structure locally:

```text
.
├── abc
│   └── __init__.py
├── collections
│   ├── __init__.py
│   ├── abc
│   │   └── __init__.py
│   └── foobar
│       └── __init__.py
├── foobar
│   ├── abc
│   │   └── __init__.py
│   └── collections
│       ├── __init__.py
│       ├── abc
│       │   └── __init__.py
│       └── foobar
│           └── __init__.py
├── ruff.toml
└── urlparse
    └── __init__.py
```

And ran my debug build of ruff:

```text
/home/brent/astral/ruff/target/debug/ruff check --no-cache --select A005  .
abc/__init__.py:1:1: A005 Module `abc` shadows a Python standard-library module
collections/__init__.py:1:1: A005 Module `collections` shadows a Python standard-library module
collections/abc/__init__.py:1:1: A005 Module `collections` shadows a Python standard-library module
collections/foobar/__init__.py:1:1: A005 Module `collections` shadows a Python standard-library module
Found 4 errors.
```

If I enable strict checking, I get 4 additional errors for the `foobar` subpackages, which is the behavior I understood from the comment.

Again, I could be missing something, but we don't really have to treat the fully-qualified path or worry about parent-child relationships mentioned in the comment directly because we check each of the files separately.

MichaReiser commented on crates/ruff_linter/src/rules/flake8_builtins/rules/stdlib_module_shadowing.rs:
```diff
@@ -60,19 +61,26 @@ impl Violation for StdlibModuleShadowing {
+    // for modules, we need to check the path grandparent in non-strict mode, not the parent of the
+    // __init__.py file
+    let (module_name, parent) = if is_module_file(path) {
+        (
+            package.path().file_name().unwrap().to_string_lossy(),
+            package.path().parent(),
+        )
     } else {
-        path.file_stem().unwrap().to_string_lossy()
+        (path.file_stem().unwrap().to_string_lossy(), path.parent())
     };
```
Nit: the only difference between the two branches seems to be whether we call the operations on `package.path` or `path`. Could we change the `if` to either return `package.path` or `path` and then extract the `module_name` and `parent` outside the `if`?

ntBre replied:
There's also `file_name` vs `file_stem` to remove the `.py` from the non-module file, but this did look a bit better before I made it a tuple. I'll think about an improvement.

MichaReiser replied:
Oh, I missed that

MichaReiser commented on crates/ruff_workspace/src/options.rs:
```diff
@@ -1143,13 +1143,21 @@ pub struct Flake8BuiltinsOptions {
     /// List of builtin module names to allow.
     pub builtins_allowed_modules: Option<Vec<String>>,
+    #[option(
+        default = r#"false"#,
+        value_type = "bool",
+        example = "builtins-strict-checking = bool"
+    )]
+    /// Compare module names instead of full module paths.
+    pub builtins_strict_checking: Option<bool>,
```
We can change the default for preview. Although I'm not sure if we did that before but it would give us some preview testing. 

The easiest for tracking is to create an issue and tag it with the next milestone. Or you can create a follow up PR, mark it as breaking, and add it to the milestone.

ntBre commented:
I think my example from earlier was actually misleading because I didn't have a `foobar/__init__.py` file. With that file, the existing behavior in ruff is already as described in the issue comment. We get the four errors above and don't error on the `foobar/` subtree. This is because the `package.path()` always seems to return the root package, rather than the leaf package. For example, when `path` is `foobar/abc/__init__.py`, the `package.path()` is still `foobar`.

Related to your `is_module_path` suggestion, I think if we want the strictness option to have any effect, we should actually check `path.file_name()` in both cases and not even use the `package` argument at all.

I noticed this when testing a default strict value and not being able to reproduce the eight-error case.

ntBre commented:
Sorry @MichaReiser, I now believe that you were right all along. I ended up writing a few integration tests with the exact module structure above (including `foobar/__init__.py`), and trying to pass those pushed me back toward the path-component implementation.

I also changed the default for `builtins_strict_checking` to `true`, so I expected the stable ecosystem check to show no changes and the preview check to match the previous report (-88), but apparently I've made the default even more strict than it was before. Is that okay, or should I try to get rid of the new violations? It looks like these are true positives, if our intention is to flag any module with a name from the standard library.

Once the non-strict behavior becomes default, there will still be a substantial net decrease in violations too.

MichaReiser commented on crates/ruff_linter/src/rules/flake8_builtins/rules/stdlib_module_shadowing.rs:
```diff
@@ -58,44 +58,78 @@ impl Violation for StdlibModuleShadowing {
+        .sorted_by_key(|s| s.as_os_str().len())
+        .rev()
+    {
+        if let Ok(new_path) = path.strip_prefix(dir) {
+            path = new_path;
+        }
```
Shouldn't we break after we found the first path or we might end up stripping the same path more than once.

MichaReiser commented:
> but apparently I've made the default even more strict than it was before. Is that okay, or should I try to get rid of the new violations? 

Can you explain in what ways you made the check more strict than it used to be before?

ntBre commented:
> Can you explain in what ways you made the check more strict than it used to be before? I ask because it depends if the changed behavior classifies as a bug fix (fine to do in a patch release) or a behavior change (it depends but generally no)

For modules, the original code used `package.path()` to get the `module_name`, which appears to give the top-level module name and I think could be considered a bug. This leads to the following output on `main`:

```text
path                                   module_name
----                                   ------------
foobar/collections/abc/__init__.py     foobar
foobar/abc/__init__.py                 foobar
collections/__init__.py                collections
abc/__init__.py                        abc
urlparse/__init__.py                   urlparse
collections/abc/__init__.py            collections
foobar/collections/__init__.py         foobar
foobar/collections/foobar/__init__.py  foobar
foobar/__init__.py                     foobar
collections/foobar/__init__.py         collections
foobar/logging.py                      logging      # new, not in tree above
```

The `package.path()` behavior is different from the treatment of `.py` files, where `main` uses the basename of the file. Instead of checking one or the other, we now consider each component of the module path separately in strict mode, so all of the `foobar/` subtree (`abc`, `collections`, `collections/foobar`) is marked with errors.

I think we're being more faithful to the `## What it does` section in the docs now:

> Checks for modules that use the same names as Python standard-library modules.

But I'm definitely wary of introducing *more* diagnostics, especially when the issue I set out to fix was for too many diagnostics.

ntBre commented:
Thanks again for the feedback here and on Discord! I think I've handled all of the review comments here and also revised the implementation to reflect the upstream behavior better. To help visualize the behavior, here's a table of the strict vs non-strict behavior, operating on the tree from above (again with `foobar/logging.py` added):

| Path                                    | Strict | Non-strict | Main | Flake8 |
|-----------------------------------------|--------|------------|------|--------|
| `abc/__init__.py`                       | ✅     | ✅         | ✅   | ✅     |
| `collections/__init__.py`               | ✅     | ✅         | ✅   | ✅     |
| `collections/abc/__init__.py`           | ✅     | ❌         | ✅   | ✅     |
| `collections/foobar/__init__.py`        | ❌     | ❌         | ✅   | ❌     |
| `foobar/__init__.py`                    | ❌     | ❌         | ❌   | ❌     |
| `foobar/abc/__init__.py`                | ✅     | ❌         | ❌   | ✅     |
| `foobar/collections/__init__.py`        | ✅     | ❌         | ❌   | ✅     |
| `foobar/collections/abc/__init__.py`    | ✅     | ❌         | ❌   | ✅     |
| `foobar/collections/foobar/__init__.py` | ❌     | ❌         | ❌   | ❌     |
| `urlparse/__init__.py`                  | ❌     | ❌         | ❌   | ❌     |
| `foobar/logging.py`                     | ✅     | ❌         | ✅   | ✅     |


Like https://github.com/astral-sh/ruff/issues/15399#issuecomment-2612005957, ✅ means we emit a diagnostic and ❌ means no diagnostic.

We now match the behavior of the upstream `flake8-builtins` in strict mode, but allow configuring a non-strict mode too.

MichaReiser commented:
Thanks for the nice table. I think we should split out the bug fix from the preview behavior change for a clear changelog entry. I hope it isn't too tedious (we can update the changelog manually if it is)

MichaReiser commented:
Exactly

MichaReiser commented:
I think I've to wait with reviewing this one until your other PR lands. I'll put it back in preview and review it when you "publish" it again.

ntBre commented:
I think this should be ready to review again. It's now only a couple of lines changed in the rule itself, the addition of the config option, and adding the right option value to a couple of existing tests.

DaniBodor commented:
I believe that the documentation for this was not yet updated, right? Was trying to read there what this new setting is without having to go through pages of discussion about it, but dont see it there.

MichaReiser commented:
The website should be updated ([see](https://docs.astral.sh/ruff/settings/#lint_flake8-builtins_builtins-strict-checking)) but yes, we should reference the option in the rule's documentation. @ntBre would you mind following up on this?

DaniBodor commented:
I indeed meant the rule documentation, which is where i (and i guess most people) normally first land when looking up how to change settings regarding a specific rule.

ntBre commented:
Will do!
